diff --git a/CMakeLists.txt b/CMakeLists.txt
index 06741a5121..ca44a83b86 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -541,7 +541,12 @@ if (UNIX)
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-implicit-function-declaration")
 endif()

-
+if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+  # These flags are needed to work around a quirk of floating point comparisons
+  # on 32-bit. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=323#c87
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fexcess-precision=standard")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fexcess-precision=standard")
+endif()

 # If Visual Studio targets are being built.
 if(MSVC)
